### ISSUE TITLE
NNI v2.7 Iteration Plan

### ISSUE THREAD

**Issue by liuzhe-lz** [microsoft/nni#4454]
# Release Plan for upcoming release
- Release manager: @J-shang
- Doc detail items: 2.9
- Feature freeze date: ~2.18~ 3.11
- Code freeze and demo date: ~2.25~
- Bugbash: 4.11
- Branch cut and next release planning date: 4.12
- Release date: ~4.13~ 4.18

### NAS (discuss on 1/17)
- [ ] P0 - Darts NAS example support (Process: training (visualize), model export, retrain, record execution results)
  - [ ] DARTS search space refactor @ dong fan
  - [ ] DARTS NAS alog refactor, and all the core NAS function @ultmaster 
  - [ ] P0 - DARTS end to end owner @JiahangXu (discussion: 1.25) pr: 2.16 (pending for ...)

NAS Search Space Hub and more - @ultmaster and dong fang DL: 2.11
- [ ]  P0 - As a user, when using NNI NAS, I can use Search Space from SearchSpaceHub on a (next step: more, ie. AutoFormer) certain stratgy (tbd by yuge) as a base search space. (能跑，跑对，灵活用）

#
- [x]  P1 - weight inheritance between multi-trial and one-shot @ultmaster
  - [x] - evaluator interface @JiahangXu DL: 2.23
- [ ]  P0 - merging layer-choices into super-kernel.

#
- [ ]  P0 - [NAS Experiment User Story][@QuanluZhang ] As an NNI user, I would like NNI to provide more user-friendly programming interface for designing and tuning DNN models, including more clear APIs, debuggability, functionality (e.g., export best model), properly combining different search modes. (design review: ~1.26~ ~2.9~ 2.14)

### Model Compression
- [ ]  P0 - transformer pruning support @J-shang  (discussion 1.18) (DL: 1.28) #4512 
- [ ]  P0 - peer comparation with competitors (internal summary first) 同比 nnPruning pruning performance, capbility, and results
- [ ] P0 - Support more modules using speedup @littlelittlehorse (pr: 1.24) #4492 
- [ ] P1 - Support pattern-based transformer pruning @linbinskn  #4481
- [ ]  P1 - rethink autoCompression https://nni.readthedocs.io/en/latest/Compression/AutoCompression.html#provide-user-model-for-compression; including: metric, scalar, schedular (discussion 1.18)
- [ ]  P1 - [Model Pruning User Story] [@J-shang ] As an NNI user, I want a configuration that can clearly describe the final pruning effect after speed up. 
- [ ]  P1 - [Model Pruning User Story] [@J-shang ] As an NNI user, in addition to pruning the model weight, I also want to prune the module input/output/activation. (discussion 1.18)
  - [ ] FLOPs Counter Refactor: support pruning v2 & more ops (attention?)
  - [ ] Config List improvement
    - [ ] Design for short term & long term
    - [ ] replace op_partial_names & exclude with lambda
  - [ ] Pruner wrapper refactor DL: 1.26 #4488 
  - [ ] Survey input/output/activation pruning algorithm
- [ ]  P1 - [Speedup User Story] [@zheng-ningxin ] As a NNI user, I would like to get a real faster NLP model after pruning for deployment. The pruned model may be from famous repo such as huggingface,  or defined by user himself. (this release, hopefully) #4149
- [ ]  P1 - [Speedup User Story] [@zheng-ningxin ] Big Accuracy drop after Pruning/speedup; 
- [ ]  P2 - [Speedup User Story] [@zheng-ningxin ] Optimize memory consumption of speedup progress, current memory is double the model size, the goal is ~ the model size;

### HPO
- [ ]  P0 - fix the critical bug of not supporting a dict as final result in v2.6 refactored tpe tuner @liuzhe-lz 
- [ ]  P0 - Bugfix: DNGO with numpy >= 1.21 crashed?
- [ ]  P2 - [HPO Tuner Suggestion User Story] @liuzhe-lz As a NNI user, with a known model in hands, I’d like NNI to provide a method to help me decide which out of box HPO Tuner is best for tuning a known model’s hyperparameters. 
- [ ]  P3 - [HPO End-to-end User Story] @liuzhe-lz  I’m new to machine learning and do not have a model. I want NNI to generate a model for my dataset fully automatic.

### Feature Engineering
- [ ] rm requirement.txt & add IT

### WebUI
- [ ] P0 - update webui doc
- [ ] P0 - update webui README(development doc) #4489
- [ ] #3746 (need backend support)  #4511
  - [ ] P1 - backend @liuzhe-lz
  - [ ] webui
- [ ] ~align issue~
- [ ] support searching customized string in the search box
- [ ] p1 - bug fix: select trial issue 
- [ ] p3 - log component(code refactor)
- [ ] kill job component(code refactor)

### Pipeline @liuzhe-lz
- [ ] P0 - Upgrade python
- [ ] p2 - cache speedup ut
- [ ] p2 - webui dev mode speedup ut
- [ ] p2 - speed up IT
- [ ] TF v1 trial -> Pytorch trial in IT @ultmaster 
- [ ] write UT or IT for search space update, import data, experiment resume @QuanluZhang 
- [ ] write UT for reporting dict as final result and intermediate result @QuanluZhang 

### Bugfix
- [ ] P1 - contrl not work @liuzhe-lz 
- [ ] trial log not work under reuse mode @littlelittlehorse 
- [ ] nni logging system refactor

### Doc Item
 - [ ] P0 - update doc some deprecated part like `builtinAssessorName` -> `name`

https://github.com/microsoft/nni/issues/4454#issuecomment-1010592199
-- 
more lower priority items in defer comments: https://github.com/microsoft/nni/issues/4454#issuecomment-1008682185

**Comment by J-shang** [comment]
### Documentation Update

## Deferred

### NAS

- [ ] P1 - @JiahangXu a TBD end2end example for NAS with Retiarii
- [ ] P1 - @QuanluZhang ~~Support tuner_no_more_trial~~
- [x] P1 - @JiahangXu review graph engine
- [ ] P? - @ultmaster Update mnasnet and example refactor (plan: #4249)
- [ ] P2 - (zebin) More choice type design
- [ ] P2 - (zebin) Survey naslib
- [ ] P2 - (zebin) Port Houwen's algorithm
- [ ] P2 - @liuzhe-lz Merge `RetiariiExperiment` back to `Experiment`
- [ ] P3 - Visualization nasbench
- [ ] P3 - Support HPO in Retiarii
- [ ] P3 - Review one-shot, refactor with lightning

### Compression

- Pruning Speed Up (check with ningxin)
- [ ] P1 - Fine-grained iterative speedup logic.
- [ ] P1 - NetAdapt Pruner
- [ ] @J-shang Transformer Pruner. (~~11.5~~ 11.8) #4180
- [ ] @zheng-ningxin Function replacement
- [ ] P1 - Sensitivity Pruner.
- [ ] @J-shang ADMM: support more metrics.
- [ ] P2 - Support min_sparsity_per_layer in global mode.
- [ ] P2 - Mixed sparse pattern.
- [ ] P2 - Support mask output/input.
- [ ] P2 - Pruning with constraints (remain 4x channels).
- [ ] P2 - @Fiascolsy Iterative pruning starts from any sparsity ratio.
- [ ] P2 - KD in pruning scheduler (maybe use original model as teacher).
- [ ] @zheng-ningxin Memory usage optimization
- [ ] Refactor of model graph generation (as an independent component)
- [ ] Speed up with special/customized layer.
- Quantization V2
- Doc
  - [ ] @J-shang Guide on how to customize pruner in v2.
  - [ ] P1 Report performance of v2 compare with paper / v1. #4365
  - [ ] Need add: fine-grained with speed up in iteration will case issue. how to write config_list.

### HPO

- [ ] P1 - fix `len(arr) > 0` problem
- [ ] P1 - (zebin) Improve hyperband & PBT with shared storage
- [ ] P3 - Multi-objective
- [ ] P2 - (zebin) transferable HPO
- [ ] P2 - Evolution nested search space

- [ ] @QuanluZhang Survey optuna, onedb, sagemaker
- [ ] @QuanluZhang fix dispatcher error, experiment status
- [ ] @ultmaster Survey ConfigSpace

### nnictl

- [ ] P2 - @J-shang Improve `nnictl experiment export`
- [ ] P2 - @liuzhe-lz auto-completion
- [ ] P2 - @liuzhe-lz early validation

### NNI Manager

- [ ] P2 - Distributed trial
- [ ] P1 - @acured Project back-end

### Web UI

- [ ] Align issue in overview page
- [ ] Support searching customized string in the search box
- [ ] Experiment list could not open experiment with prefix https://github.com/microsoft/nni/issues/3746 (need backend)

### Other

- readthedocs dependency
- doc translation

**Comment by liuzhe-lz** [comment]
## Doc Items ##

- Theme @ultmaster P0
- Different homepage for readthedocs and GitHub - P1
- File name convention
  - URL redirection plugin @liuzhe-lz P0
- Directory hierachy
  - Design
- Static and image files
  - Need contributor guide
- Translation mechanism
  - contributor guide

- Model compression doc hierachy
  - Design @J-shang
- Update model compression quick start @J-shang
- Common style for each algorithm
- List what layers each algorithm supports

- Doc for introduction to the topic @yuqing

- Need Contributor guide

**Comment by Lijiaoa** [comment]
**+1 webui related task**

- [ ] update webui doc
- [ ] update webui README(development doc)
- [ ] ~align issue~
- [ ] support searching customized string in the search box
- [ ]  #3746 
    - [ ]  backend
    - [ ] webui
- [ ] bug fix: select trial issue
- [ ] log component(code refactor)
- [ ] kill job component(code refactor)

**Comment by J-shang** [comment]
Compression

- Overview (copy paste & update outdate information)
- Pruning (basic concept)
  - Quickstart (write tutorial)
  - Pruner Selection Guide (report result & pruner suggustion)
  - Pruner Referrence (move pruner introduction to code comments)
  - Speed Up
- Quantization (basic concept)
  - Quickstart (write tutorial)
  - Quantizer Selection Guide (report result & pruner suggustion)
  - Quantizer Referrence (move quantizer introduction to code comments)
  - Speed Up
- Utilities
- Refference (config_list key & advance usage)

**Comment by ultmaster** [comment]
The following notebooks need update:

```
/home/vsts/work/1/s/docs/source/Compression/QuickStart.rst:4: WARNING: toctree contains reference to nonexisting document 'Compression/compression_pipeline_example'
/home/vsts/work/1/s/docs/source/NAS/Benchmarks.rst:4: WARNING: toctree contains reference to nonexisting document 'NAS/BenchmarksExample'
/home/vsts/work/1/s/docs/source/Tutorial/HowToLaunchFromPython.rst:4: WARNING: toctree contains reference to nonexisting document 'Tutorial/python_api_start'
/home/vsts/work/1/s/docs/source/Tutorial/HowToLaunchFromPython.rst:4: WARNING: toctree contains reference to nonexisting document 'Tutorial/python_api_connect'
```

**Comment by J-shang** [comment]
HPO

- [ ] (python API) * (TF + Free) + nnictl @liuzhe-lz 
- [ ] Pytorch @J-shang
- [ ] nnictl reference @ultmaster 
- [ ] Evolution @zebin
- [ ] SMAC @QuanluZhang 
- [ ] Metis @littlelittlehorse 
- [ ] GP @QuanluZhang 
- [ ] Hyperband @QuanluZhang 
- [ ] BOHB @QuanluZhang 
- [ ] PBT @ultmaster
- [ ] DNGO @ultmaster

**Comment by J-shang** [comment]
- [ ] API Reference TOC

**Comment by J-shang** [comment]
## Bug Bash
### Doc
- [ ] #4489 Owner: @Lijiaoa 
- [ ] #4500 Owner: @ultmaster 
- [ ] #4522 Owner: @ultmaster 
- [ ] #4513 Owner: @Lijiaoa 
- [ ] #4562 Owner: @Lijiaoa 
- [ ] #4561 Owner: @ultmaster 
- [ ] #4557 Owner: @ultmaster 
- [ ] #4573 Owner: @ultmaster 
- [ ] #4577 Owner: @QuanluZhang 
- [ ] #4584 Owner: @ultmaster 
- [ ] #4574 Owner: @J-shang 
- [ ] #4587 Owner: @ultmaster 
- [ ] #4570 Owner: @Lijiaoa 
- [ ] #4612 Owner: @QuanluZhang 
- [ ] #4616 Owner: @Lijiaoa 
- [ ] #4615 Owner: @zzpmiracle 
- [ ] #4618 Owner: @ultmaster 
- [ ] #4611 Owner: @LikeJulia 
- [ ] #4623 Owner: @J-shang 
- [ ] #4579 Owner: @liuzhe-lz 
- [ ] #4613 Owner: @ultmaster 
- [ ] #4624 Owner: @liuzhe-lz 
- [ ] #4617 Owner: @QuanluZhang 
- [ ] #4627 Owner: @ultmaster 
- [ ] #4639 Owner: @J-shang 
- [ ] #4642 Owner: @ultmaster 
- [ ] #4646 Owner: @J-shang 
- [ ] #4649 Owner: @J-shang 
- [ ] #4634 Owner: @liuzhe-lz 
- [ ] #4653 Owner: @liuzhe-lz 
- [ ] #4630 Owner: @liuzhe-lz 
- [ ] #4659 Owner: @ultmaster 
- [ ] #4589 Owner: @ultmaster 
- [ ] #4656 Owner: @liuzhe-lz 
- [ ] #4661 Owner: @liuzhe-lz 
- [ ] #4640 Owner: @liuzhe-lz 
- [ ] #4638 Owner: @liuzhe-lz 
- [ ] #4575 Owner: @ChrisZRen
- [ ] #4667 Owner: @J-shang 
- [ ] #4672 Owner: @liuzhe-lz 
- [ ] #4560 Owner: @JiahangXu 
- [ ] #4677 Owner: @ultmaster 
- [ ] #4680 Owner: @Lijiaoa 
### NAS
- [ ] #4473 Owner: @ultmaster 
- [ ] #4399 Owner: @ultmaster 
- [ ] #4411 Owner: @ultmaster 
- [ ] #4435 Owner: @ultmaster 
- [ ] #4479 Owner: @ultmaster 
- [ ] #4586 Owner: @ultmaster 
- [ ] #4544 Owner: @ultmaster 
- [ ] #4595 Owner: @JiahangXu 
- [ ] #4622 Owner: @ultmaster 
- [ ] #4600 Owner: @congxu-ml 
### Compression
- [ ] #4481 Owner: @linbinskn 
- [ ] #4534 Owner: @BasicCoder 
- [ ] #4628 Owner: @zheng-ningxin 
### HPO
### NNI Manager
- [ ] #4486 Owner: @liuzhe-lz 
- [ ] #4506 Owner: @weidankong 
- [ ] #4510 Owner: @liuzhe-lz 
### WebUI
- [ ] #4496 Owner: @liuzhe-lz 
- [ ] #4535 Owner: @liuzhe-lz 
- [ ] #4548 Owner: @Lijiaoa 
### Bug Fix
- [ ] #4511 Owner: @Lijiaoa 
- [ ] #4487 Owner: @ultmaster 
- [ ] #4520 Owner: @mrshu 
- [ ] #4531 Owner: @liuzhe-lz 
- [ ] #4541 Owner: @Lijiaoa 
- [ ] #4593 Owner: @weidankong 
- [ ] #4552 Owner: @ultmaster 
- [ ] #4594 Owner: @zheng-ningxin 
- [ ] #4614 Owner: @JiahangXu 
- [ ] #4652 Owner: @uniartisan 
- [ ] #4633 Owner: @weidankong 
### Pipelines
- [ ] #4475 Owner: @liuzhe-lz 
- [ ] #4495Owner: @liuzhe-lz 
- [ ] #4503 Owner: @liuzhe-lz 
- [ ] #4558 Owner: @J-shang 
- [ ] #4580 Owner: @J-shang 
- [ ] #4569 Owner: @J-shang 
### Others
- [ ] #4507 Owner: @a1trl9 
- [ ] #4476 Owner: @liuzhe-lz 
- [ ] #4534 Owner: @fdiblen 
- [ ] #4576 Owner: @liuzhe-lz

**Comment by J-shang** [comment]
- [ ] #4696 
- [ ] #4644
- [ ] #4662
- [ ] #4691
